### PR TITLE
Remove Tensor.h, TensorMethods.h from src/core.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27086 Remove Tensor.h, TensorMethods.h from src/core.**

Checked in generated files are a major source of merge conflicts. This is a commandeer of #25031.

Unfortunately, this patch is pretty involved. The general concept is to excise Caffe's dependency on ATen/core entirely. In practice, this means eliminating includes of `ATen/core/Tensor.h` and also all occurrences of IValues (who mention Tensor in their public interface.) This comes in several steps:

1. Modify all existing `CAFFE2_IS_XPLAT_BUILD` ifdefs to also consider `C10_MOBILE`. So basically, bring the OSS Caffe2 build and the XPLAT build in line. That eliminates a bunch of dependencies on TensorMethods.h. However, it's not enough, we need to...
2. Delete "c10 experimental" from the mobile build (cc @smessmer ) as empirically they are built as part of caffe2 mobile. We also need to
3. Remove a few more occurrences of IValue that don't seem to be handled by xplat today, e.g., caffe2/operators/reduction_ops.h (this is probably because we don't actually ship these ops in xplat today)
4. Cleanup on GradMode (ifdef'ed out!) and Context registration (move that header/cpp file back to Caffe2)

Differential Revision: [D17687345](https://our.internmc.facebook.com/intern/diff/D17687345)

### CODE REVIEW

**Comment by ezyang** [comment]
I believe the plan on record for Caffe2 Android build is that it should not need any deps on ATen/core

File: caffe2/core/context_base.cc [hunk 1]
```diff
@@ -1,5 +1,20 @@
-#include "context_base.h"
+#include <caffe2/core/context_base.h>
```
Comment (ezyang) [inline caffe2/core/context_base.cc]: Code movement only (except for header locations update)

File: caffe2/core/context_base.h [hunk 2]
```diff
@@ -1,7 +1,168 @@
...
+#include <c10/macros/Macros.h>
+#include <c10/core/Allocator.h>
+#include <c10/util/typeid.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Registry.h>
+#include <c10/core/CopyBytes.h>
+
 #include "caffe2/core/common.h"
 #include "caffe2/core/logging.h"
 #include "caffe2/proto/caffe2_pb.h"
```
Comment (ezyang) [inline caffe2/core/context_base.h]: Code movement only, except the headers that were previously in `caffe/core/context_base.h` are now inlined directly in the file.

File: caffe2/core/export_c10_op_to_caffe2.h [hunk 3]
```diff
@@ -1,13 +1,16 @@
 #pragma once
 
+#include <c10/macros/Macros.h>
+#include <c10/util/Registry.h>
+#include "caffe2/core/operator.h"
```
Comment (ezyang) [inline caffe2/core/export_c10_op_to_caffe2.h]: Some downstream deps (not in xplat, apparently?) seem to rely on this being in scope, so I whisked it out of the ifdef.

File: caffe2/CMakeLists.txt [hunk 4]
```diff
@@ -5,6 +5,7 @@ include(../cmake/Codegen.cmake)
 
 # ---[ ATen build
 if (INTERN_BUILD_ATEN_OPS)
+  message(WARNING "NOT only building core")
```
Comment (zou3519) [inline caffe2/CMakeLists.txt]: What does it mean to "NOT only [build] core"?

File: caffe2/core/export_c10_op_to_caffe2.h [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
+
 // TODO Also register c10 operators on mobile
-#if !defined(CAFFE2_IS_XPLAT_BUILD)
+#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```
Comment (zou3519) [inline caffe2/core/export_c10_op_to_caffe2.h]: Are there instances where we only consider CAFFE2_IS_XPLAT_BUIILD or C10_MOBILE? Could we delete one of these two in the future?

I remember @ljk53 was talking about consolidating the mobile build flags.

**Review by zou3519 — COMMENTED** [review]
(not full review) why do we have to excise caffe2's dependency on aten/core to delete TensorMethods.h and TensorBody.h? Is this to clean up the code or does caffe2 not do codegen?

**Comment by ezyang** [comment]
> why do we have to excise caffe2's dependency on aten/core to delete TensorMethods.h and TensorBody.h? Is this to clean up the code or does caffe2 not do codegen?

Good question. There are multiple viable approaches. This particular approach has the benefit that I don't  have to setup codegen for OSS Android build, or in Caffe2 xplat. The alternative is to not excise, and instead teach Caffe2 Android and xplat build how to do codegen. This seemed hard (to me, anyway), so I decided not to do it.

File: aten/src/ATen/gen.py [hunk 6]
```diff
@@ -45,12 +45,7 @@
     action='store_true',
     help='reinterpret CUDA as ROCm/HIP and adjust filepaths accordingly')
 options = parser.parse_args()
-gen_to_source = os.environ.get('GEN_TO_SOURCE')  # update source directly as part of gen
-if not gen_to_source:
-    core_install_dir = os.path.join(options.install_dir, 'core_tmp') if options.install_dir is not None else None
-else:
-    core_install_dir = os.path.join(options.source_path, 'core')
-
+core_install_dir = options.install_dir + '/core' if options.install_dir is not None else None
```
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/gen.py]: we should use os.path.join here

File: caffe2/CMakeLists.txt (hunk 4 above)
Comment (ezyang) [reply caffe2/CMakeLists.txt]: This is old from greg's patch, it can be deleted

File: caffe2/core/export_c10_op_to_caffe2.h (hunk 5 above)
Comment (ezyang) [reply caffe2/core/export_c10_op_to_caffe2.h]: It looks like the answer is that, most of the time, we consider both as a pair. The only place this is not true is:

```
aten/src/ATen/core/op_registration/op_registration.cpp
2:#if !defined(CAFFE2_IS_XPLAT_BUILD)
                                                              
aten/src/ATen/core/op_registration/op_registration.h          
10:#if !defined(CAFFE2_IS_XPLAT_BUILD)                                      
99:        #if defined(CAFFE2_IS_XPLAT_BUILD)

aten/src/ATen/core/EnableNamedTensor.h
9:#if !defined(CAFFE2_IS_XPLAT_BUILD) && (!defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE))
```

So agreed here.

**Review by zou3519 — APPROVED** [review]
lgtm, but I don't have context on the caffe2 bits. The guards should be as fine as all the builds build.

File: aten/src/ATen/gen.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/gen.py]: Actually, apparently, NOT using os.path.join is mandatory on Windows. With this change, you get:

```

15:15:11 CMake Error at caffe2/CMakeLists.txt:600 (add_library):
15:15:11   Syntax error in cmake code when parsing string
15:15:11 
15:15:11     C:/Jenkins/workspace/pytorch-builds/pytorch-win-ws2016-cuda9-cudnn7-py3-build/build/aten/src/ATen\core/TensorMethods.h
15:15:11 
15:15:11   Invalid character escape '\c'.
```

[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/gen.py]: That's... counterintuitive, but go for it. How did this work before?

[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/gen.py]: ya I don't see how this worked before, then.

**Review by gchanan — APPROVED** [review]

**Comment by gchanan** [comment]
this seems kind of annoying to merge.  Perhaps you should break it up into:
1) separate caffe2 bits
2) codegen changes

You could also consider temporarily keeping the checked in files but ignoring them (so we don't break in-flight patches), but that might not be worth it.

**Comment by ljk53** [comment]
@ezyang I updated internal diff D16983025 to change fbcode to use the generated files. That diff depends on some change on this PR (core_tmp -> core) and this PR needs that diff to not break fbcode. We need land them together or you can merge my change into D17687345 or maybe you should consider keeping generated files for a while as suggested by @gchanan then after D16983025 is landed we can remove generated files.

File: caffe2/core/export_c10_op_to_caffe2.h (hunk 5 above)
Comment (ljk53) [reply caffe2/core/export_c10_op_to_caffe2.h]: I think semantically CAFFE2_IS_XPLAT_BUILD should be equivalent to "C10_MOBILE && ! FEATURE_TORCH_MOBILE" - it's xplat mobile build (thus C10_MOBILE) and caffe2 only as in current production (thus !FEATURE_TORCH_MOBILE). 

One argument @smessmer had to keep CAFFE2_IS_XPLAT_BUILD was that xplat has broader meaning than mobile, but I think the substantial difference is whether to build full torch (FEATURE_TORCH_MOBILE).

**Comment by ezyang** [comment]
> this seems kind of annoying to merge

Does that mean if I manage to merge it, all is well? :) Unlike other patches, this won't silently merge conflict with anyone else's stuff. I might just do it over the weekend.

@ljk53 I'll just squash the diffs together.

File: caffe2/core/export_c10_op_to_caffe2.h (hunk 5 above)
Comment (ezyang) [reply caffe2/core/export_c10_op_to_caffe2.h]: I don't really have a stone in the game, except that I don't want to make too many more changes to this diff as it stands right now :)

**Pull request closed without merging**